source/iot/MqttRequestResponseClient.cpp

```diff
@@ -0,0 +1,509 @@
+                    }
+                }
+
+                ~StreamReadLock()
+                {
+                    if (m_taken)
+                    {
+                        aws_rw_lock_runlock(m_lock);
+                    }
+                }
+
+              private:
+                struct aws_rw_lock *m_lock;
+
+                bool m_taken;
+            };
+
+            class StreamWriteLock
+            {
+              public:
+                StreamWriteLock(struct aws_rw_lock *lock, struct aws_event_loop *protocolLoop)
+                    : m_lock(lock), m_taken(false)
+                {
+                    if (!aws_event_loop_thread_is_callers_thread(protocolLoop))
+                    {
+                        m_taken = true;
+                        aws_rw_lock_wlock(lock);
+                    }
+                }
+
+                ~StreamWriteLock()
+                {
+                    if (m_taken)
+                    {
+                        aws_rw_lock_wunlock(m_lock);
+                    }
+                }
+
+              private:
+                struct aws_rw_lock *m_lock;
+
+                bool m_taken;
+            };
+
+            class StreamingOperationImpl
+            {
+              public:
+                StreamingOperationImpl(
+                    struct aws_mqtt_rr_client_operation *stream,
+                    const StreamingOperationOptionsInternal &options,
+                    struct aws_event_loop *protocolLoop);
+                virtual ~StreamingOperationImpl();
+
+                void Open();
+
+                void Close();
+
+                static void OnSubscriptionStatusCallback(
+                    enum aws_rr_streaming_subscription_event_type status,
+                    int error_code,
+                    void *user_data);
+
+                static void OnIncomingPublishCallback(struct aws_byte_cursor payload, void *user_data);
+
+                static void OnTerminatedCallback(void *user_data);
+
+              private:
+                StreamingOperationOptionsInternal m_config;
+
+                struct aws_mqtt_rr_client_operation *m_stream;
+
+                struct aws_event_loop *m_protocolLoop;
+
+                struct aws_rw_lock m_lock;
+
+                bool m_closed;
+            };
+
+            struct StreamingOperationImplHandle
+            {
+                StreamingOperationImplHandle() : m_allocator(nullptr), m_impl(nullptr) {}
+
+                Aws::Crt::Allocator *m_allocator;
+
+                std::shared_ptr<StreamingOperationImpl> m_impl;
+            };
+
+            StreamingOperationImpl::StreamingOperationImpl(
+                struct aws_mqtt_rr_client_operation *stream,
+                const StreamingOperationOptionsInternal &options,
+                struct aws_event_loop *protocolLoop)
+                : m_config(options), m_stream(stream), m_protocolLoop(protocolLoop), m_lock(), m_closed(false)
```

Comment:
nit: Maybe validate that `options` has all fields set? Because down the code we execute `m_config` members without checking them.

Reply:
Added function containment checks before the callbacks